internal/pipeline/executor.go

```diff
@@ -0,0 +1,290 @@
+			// already hydrated, we sort of assume pkgs in hctx will be populated but we
+			// can be defensive here
+			resources = hctx.pkgs[p.Path()].resources
+		default:
+			err = fmt.Errorf("invalid pkg state %v detected", state)
+		}
+		return resources, err
+	}
+	// mark the pkg in hydrating
+	hctx.hydrationStates[p.Path()] = Hydrating
+	// add to the graph
+	hctx.pkgs[p.Path()] = p
+
+	// read the pkg, the pipeline, dependencies etc.
+	err = p.Read()
+	if err != nil {
+		err = fmt.Errorf("failed to read the pkg %w", err)
+		return resources, err
+	}
+
+	// input resources for current pkg's hydration
+	var input []*yaml.RNode
+	// here we explore the p.sources to determine sub packages etc.
+	for _, s := range p.resolveSources() {
+		// TODO(droot): sync with pipeline's defaults once Donny's PR is merged
+		if s == p.Path() {
+			var currPkgResources []*yaml.RNode
+			// current pkg will be hydrated as we drop of this loop
+			currPkgResources, err = p.localResources(false)
+			if err != nil {
+				return resources, err
+			}
+			input = append(input, currPkgResources...)
+		} else {
+			// TODO(droot): recursive hydration is broken for now, needs more work
+			var depResources []*yaml.RNode
+			var depPkg *pkg
+
+			if depPkg, err = newPkg(s); err != nil {
+				return resources, err
+			}
+
+			depResources, err = hydrate(depPkg, hctx)
+			if err != nil {
+				err = fmt.Errorf("failed to hydrated sub pkg %s %w", s, err)
+				return resources, err
+			}
+			input = append(input, depResources...)
+		}
+	}
+
+	output := &kio.PackageBuffer{}
+	// TODO(droot): parameterize the sink-mode (hctx.sinkMode) to determine whether to
+	// write it in-place or not. Currently in-place.
+	pkgWriter := &kio.LocalPackageWriter{PackagePath: p.Path()}
+	// create a kio pipeline from kyaml library to execute the function chains
+	kioPipeline := kio.Pipeline{
+		Inputs: []kio.Reader{
+			&kio.PackageBuffer{Nodes: input},
+		},
+		Filters: fnFilters(p.Pipeline()),         // we will gather filters from the pipeline
+		Outputs: []kio.Writer{pkgWriter, output}, // here may be we don't want to write to the fs yet
+	}
+	err = kioPipeline.Execute()
+	if err != nil {
+		err = fmt.Errorf("failed to run pipeline for pkg %s %w", p.path, err)
+		return
+	}
+
+	// pkg is hydrated
+	resources = output.Nodes
+	// mark the pkg as wet and update the resources
+	hctx.hydrationStates[p.Path()] = Wet
+	p.resources = resources
+	return resources, err
+}
+
+// TODO(droot): filter kpt metadata files such as pipeline, Kptfile, permissions etc.
+func filterMetaData(resources []*yaml.RNode) []*yaml.RNode {
+	return resources
+}
+
+// fnFilters returns chain of functions that are applicable
+// to a given pipeline.
+func fnFilters(_ *Pipeline) []kio.Filter {
+	// TODO(droot): hardcoding built-in set-annotation function for testing
+	// this will be replaced from the pipeline functions eventually.
+	fn := &annotator{
+		key:   "builtin/setter-1",
+		value: "test",
+	}
+
+	return []kio.Filter{
```

Comment:
Maybe another TODO here for adding other functions from real pipeline.

Reply:
Done.